037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,261 @@
+The only fields that the cluster operator has to specify, when creating the `KafkaRebalance` custom resource, based on scaling up or down, are the:
+
+* `spec.mode` : should be `add-brokers` or `remove-brokers` if scaling up or down.
+* `spec.brokers` : with the list of brokers to add/remove.
+* `spec.autoApproval` : for allowing the auto-approval as defined by this [proposal](https://github.com/strimzi/proposals/pull/56).
+
+The drawback of this solution is that if the user wants to customize the rebalancing, he has to use the `spec.cruiseControl.config` field in the Kafka custom resource. It’s important to mention that it affects the overall Cruise Control configuration so that any other “empty” rebalancing request, not related to scaling, will use the same options.
+Users won't also be able to skip hard goals to move forward with a rebalancing if they are not met anyway.
+
+To allow customization, a potential solution would be to add rebalancing related configuration (goals, and so on) under the `spec.cruiseControl.autoRebalance` directly.
+It would be a kind of copy of the fields from the `KafkaRebalance` custom resource excluding the mode and brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    replicas: 3
+    # ...
+  cruiseControl:
+    # ...  
+    autoRebalance:
+      # using the same rebalancing configuration when adding or removing brokers ... a lot of duplication
+      - mode: add-brokers
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+          - RackAwareGoal
+          - MinTopicLeadersPerBrokerGoal
+          - NetworkOutboundCapacityGoal
+          - ReplicaCapacityGoal
+        skipHardGoalCheck: true
+        # ... other rebalancing related configuration
+      - mode: remove-brokers
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+          - RackAwareGoal
+          - MinTopicLeadersPerBrokerGoal
+          - NetworkOutboundCapacityGoal
+          - ReplicaCapacityGoal
+        skipHardGoalCheck: true
+        # ... other rebalancing related configuration  
+      - mode: full
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+        skipHardGoalCheck: false
+        # ... other rebalancing related configuration  
+```
+
+For handling the auto-approval, a `spec.autoApproval` field, as described by this [proposal](https://github.com/strimzi/proposals/pull/56), would be needed as well.
+This proposal was rejected because it pollutes the `Kafka` custom resource with too many details and configuration related to a rebalancing operation when the user wants to specify specific desires compared to the default ones in the Cruise Control configuration.
```

Comment:
I really like this proposal of a `spec.cruiseControl.autoRebalance` field! Is it already too late to discuss this proposal further as an option, was already rejected by everyone?? 

I understand it would add a lot of new fields to the `Kafka` resource but I think it would still be easier for to develop and use!

Reply:
It's not late for anything but what I see with this proposal is that the `Kafka` custom resource can really be poisoned with a lot of rebalancing related stuff becoming bigger and bigger even with a lot of duplication.

Reply:
> Kafka custom resource can really be poisoned with a lot of rebalancing related stuff becoming bigger and bigger even with a lot of duplication.

It would definitely bloat the Kafka resource but I feel it would be more intuitive of an interface and less complex to implement than using another custom resource especially one so similar to the existing `KafkaRebalance` resource. 

@katheris @scholzj @tomncooper I assume since this rejected alternative was not commented on that you agree that this route is not worth bloating the Kafka resource, is that correct?

Reply:
I am still unsure about the use of both a `KafkaRebalance` and `KafkaRebalanceTemplate` where the `KafkaRebalance` could be owned by the user or could be owned by Strimzi, however I don't think adding it to the `Kafka` resource makes it simpler. I think it is nice to be able to handle rebalance configs in a similar way between the auto-approved ones and the manual approval ones. My worry is more with the different types of `KafkaRebalance` than adding a new CR.

Reply:
I think we are moving around in circle on this ... 
Having a new `KafkaRebalanceTemplate` CR seems to be much but allows the flexibility to specify rebalancing configuration, but then it comes to user created `KafkaRebalance` vs operator created `KafkaRebalance`.
Bloating the `Kafka` CR doesn't sound great to me unless we decide to move the entire `cruiseControl` related configuration in a new `KafkaCruiseControl` custom resource which will have the current CC configuration and it could host the auto-rebalancing as well. In that case, I would agree that a new `KafkaRebalanceTemplate` wouldn't be necessary (by "bloating the new `KafkaCruiseControl` resource this time). Anyway, it would not "fix" the "issue" to have both user created `KafkaRebalance` vs operator created `KafkaRebalance` because trying to leverage the state machine we already have in place.
Of course, it would be something t   o do before moving forward with this PR. It would be a pre-requisite somehow.
@katheris @kyguy @scholzj @tombentley @tomncooper should we have a thought about this first? Wdyt?

Reply:
@ppatierno `KafkaCruiseControl` might not be a bad idea, but if you want to do that, it would probably be for a separate proposal which should be dealt with before this. So not sure how eager you are to go there. My main confusion here is why do we need so many different goals definitions. But that might very well be because of my lack of CC knowledge. My main `KafkaRebalanceTemplate` concern is that it means a lot of maintenance effort going forward. But if you think it is needed despite the effort, then you should go for it -> I do not have problem to approve the PR with it.

Reply:
> KafkaCruiseControl might not be a bad idea, but if you want to do that, it would probably be for a separate proposal which should be dealt with before this. So not sure how eager you are to go there.

Well let's say that apart from the user goals customization, moving to a `KafkaCruiseControl` would leave us more freedom to approach to the problem without taking the `Kafka` CR anymore and not being scared about polluting it if needed. I am fine to work on a different proposal if we all think that will be helpful anyway. Of course, we'll park this one for later.

Reply:
It would be a big change, so I think wider concensus is needed (i.e. we need more people to comment apart from the two of us ;-))

Reply:
For the `spec.cruiseControl.autoRebalance` argument but also applies to the proposed `KafkaCruiseControl` as well:

Would we have to include the `goal` and other rebalance options [1] as configurable through the Kafka resource? Since the Cluster Operator is the client executing the rebalance call couldn't the Cluster Operator be responsible for determining/setting the optimal rebalance options. [1] I If the Cluster Operator is responsible for those rebalance options [1] then we don't need to expose those rebalance options to users through the Kafka custom resource and we would save the Kafka custom resource from the bloat! I think it makes sense to have the autoscale up and down to rely on the default goals in the `spec.cruiseControl.config` anyway since the overall cluster balancedness score will be based on those `spec.cruiseControl.config` configured goals  before and after the scaling, right? What's the use of the `spec.cruiseControl.config` goals settings if we are overriding them all the time anyway?

[1] https://github.com/linkedin/cruise-control/wiki/REST-APIs#add-a-list-of-new-brokers-to-kafka-cluster